Beam/Include/Beam/Queries/AsyncDataStore.hpp

```diff
@@ -0,0 +1,216 @@
+      auto additionalMatches = std::vector<SequencedValue>();
+      if(query.GetSnapshotLimit().GetType() == SnapshotLimit::Type::HEAD) {
+        additionalMatches = buffer->Load(query);
+      } else {
+        additionalMatches = m_dataStore->Load(query);
+      }
+      auto mergedMatches = std::vector<SequencedValue>();
+      MergeWithoutDuplicates(additionalMatches.begin(), additionalMatches.end(),
+        matches.begin(), matches.end(), std::back_inserter(mergedMatches),
+        SequenceComparator());
+      if(static_cast<int>(mergedMatches.size()) >
+          query.GetSnapshotLimit().GetSize()) {
+        if(query.GetSnapshotLimit().GetType() == SnapshotLimit::Type::HEAD) {
+          mergedMatches.erase(mergedMatches.begin() +
+            query.GetSnapshotLimit().GetSize(), mergedMatches.end());
+        } else {
+          mergedMatches.erase(mergedMatches.begin(),
+            mergedMatches.begin() +
+            (mergedMatches.size() - query.GetSnapshotLimit().GetSize()));
+        }
+      }
+      matches = std::move(mergedMatches);
+    }
+    return matches;
+  }
+
+  template<typename D, typename E>
+  void AsyncDataStore<D, E>::Store(const IndexedValue& value) {
+    auto lock = boost::lock_guard(m_mutex);
+    m_currentDataStore->Store(value);
+    TestFlush();
+  }
+
+  template<typename D, typename E>
+  void AsyncDataStore<D, E>::Store(const std::vector<IndexedValue>& values) {
+    auto lock = boost::lock_guard(m_mutex);
+    m_currentDataStore->Store(values);
+    TestFlush();
+  }
+
+  template<typename D, typename E>
+  AsyncDataStore<D, E>::~AsyncDataStore() {
+    Close();
+  }
+
+  template<typename D, typename E>
+  void AsyncDataStore<D, E>::Open() {
+    if(m_openState.SetOpening()) {
+      return;
+    }
+    try {
+      m_currentDataStore->Open();
+      m_dataStore->Open();
+    } catch(const std::exception&) {
+      m_openState.SetOpenFailure();
+      Shutdown();
+    }
+    m_openState.SetOpen();
+  }
+
+  template<typename D, typename E>
+  void AsyncDataStore<D, E>::Close() {
+    if(m_openState.SetClosing()) {
+      return;
+    }
+    Shutdown();
+  }
+
+  template<typename D, typename E>
+  void AsyncDataStore<D, E>::Shutdown() {
+    auto writeToken = Routines::Async<void>();
+    m_tasks.Push(
+      [&] {
+        Flush();
+        writeToken.GetEval().SetResult();
+      });
+    writeToken.Get();
+    m_openState.SetClosed();
+  }
+
+  template<typename D, typename E>
+  void AsyncDataStore<D, E>::TestFlush() {
+    if(m_flushPending) {
+      return;
+    }
+    m_flushPending = true;
+    m_tasks.Push(
+      [=] {
+      Flush();
+    });
+  }
+
+  template<typename D, typename E>
+  void AsyncDataStore<D, E>::Flush() {
+    auto dataStore = std::make_shared<ReserveDataStore>();
+    {
+      auto lock = boost::lock_guard(m_mutex);
+      dataStore.swap(m_currentDataStore);
+      m_flushPending = false;
+    }
+    m_dataStore->Store(dataStore->LoadAll());
```

Comment:
The issue as I see it is that if an ```AsyncDataStore::Load``` happens from another thread while the current thread is executing line 208, then none of the data in ```dataStore``` will be returned. Can you correct this bug in ```AsyncDataStore```? I can then apply the fix to ```BufferedDataStore```.

Reply:
I don't quite get why. Load reads data from `m_flushedDataStore` in addition to `m_dataStore`. `m_flushedDataStore` is normally bound to `m_currentDataStore` (they are 2 copies of the same shared pointers), but on line 205, `m_currentDataStore` gets temporarily detached from from `m_flushedDataStore` because it is replaced with a new local data store to allow for new writes during the flush process. `m_flushedDataStore` still points to the old data store, up to line 211, which is now being flushed using the `dataStore` handle. As a result, `Load` will retrieve all entries from `m_dataStore` and `dataStore`, removing possible duplicates. Am I missing something?

Reply:
Okay, then Load needs to load from m_dataStore to get all persisted data... m_flushedDataStore to get all data currently being flushed, and m_currentDataStore to get any data that has not been flushed before and is not currently being flushed, correct?

Reply:
Okay, I think I am getting that now. If the flush process has started but not yet finished, new Stores occur followed by Loads, Loads won't see data from those new Stores as they are writing to `m_currentDataStore` not yet bound to `m_flushedDataStore`.
Reading from all three data store handled does not seem promising performance-wise because simultaneously reading from and writing to `m_currentDataStore` would require locking excessive locking.
I can, however, try to only trigger the heavy `m_currentDataStore` read only when there is a flush process in progress. That can be published using another flag, which is checked inside the critical section in `Load`. If the flush is happening at that time, all entries are loaded from `m_currentDataStore` while the critical section is still open. A new flush process can begin anywhere after the critical section ends, but it does not matter as only new Stores (called after Load was called) will possibly be unaccounted for.

Reply:
Should we follow the approach described above, we should probably use a shared lock as loaders may get heavily interlocked during `m_currentDataStore` loads.

Reply:
As I mention in my other comment, feel free to take some liberties to implement a solution as you see fit, given the time constraints and implementation requirements. You can always submit an initial implementation now that is reasonably fast, and then work on further optimizations later.